Fix Battery Parameter Migration

MaEtUgR opened the pull request:
**Describe problem solved by this pull request**
@cmic0 tried to change `BAT1_V_EMPTY`. Parameter was changed, it differed from `BAT_V_EMPTY` and after a reboot (starting simulation again) it was back to the old value.

Battery parameter migration from `BAT_*` to `BAT1_*` had the following issues:
1. In simulation the `SimulatorBattery` broke the chain of ModuleParams parent/child tree which the logic is relying on. Additionally it overwrote the parameters with hardcoded values internally.
1. The parameter migration prefers to migrate old over new if the values differ on boot but the flag `_first_parameter_update` was initialized false and hence always false. This didn't have an effect most of the time because the `previous_old_val` is zero on initialization and if the new parameter isn't zero it went the right path.
1. Some drivers using the battery class e.g. INA266 didn't clear the `parameter_update` subscription once they checked it's updated.
1. Some parts were hard to read.

**Describe your solution**
1. Remove `SimulatorBattery`, it only did harm.
1. Initialize `_first_parameter_update` with true.
1. Copy subscription data to a dummy to reset after it was updated.
1. Refactoring.

**Test data / coverage**
SITL testing.

dagar commented on src/drivers/power_monitor/voxlpm/voxlpm.cpp:
```diff
@@ -318,7 +318,11 @@ VOXLPM::measure()
 
 	parameter_update_s update;
```
delete?

MaEtUgR replied:
Forgot it there 🤦‍♂️

dagar commented on src/lib/battery/battery.cpp:
```diff
@@ -186,7 +186,8 @@ Battery::updateBatteryStatus(hrt_abstime timestamp, float voltage_v, float curre
 {
-	orb_publish_auto(ORB_ID(battery_status), &_orb_advert, &_battery_status, &_orb_instance);
+	int dummy; // We're not interested in the instance ID we get
+	orb_publish_auto(ORB_ID(battery_status), &_battery_status_pub, &_battery_status, &dummy);
```
If you're going to change this how about replacing it with a `uORB::PublicationMulti<battery_status_s>`?

MaEtUgR replied:
I thought about that but don't know what's the equivalent of this auto publication. Don't you have to specify the ID in advance? Would you just use the `const int _index;` of the battery for that?

MaEtUgR replied:
I tried to add it. Works with one battery (SITL). Is that the correct use? https://github.com/PX4/Firmware/pull/15616/commits/a153aa5aa66300427e06c6fd2dcd9f3ab5e34f74
Does it automatically choose the next free instance?

dagar commented:
Nice work @MaEtUgR. The battery param migration was a real mess that I'm looking forward to purging once the QGC side is in place.

dagar commented on ROMFS/px4fmu_common/init.d-posix/rcS:
```diff
@@ -119,7 +119,7 @@ if [ $AUTOCNF = yes ]
 	param set SYS_AUTOSTART $REQUESTED_AUTOSTART
 
-	param set BAT_N_CELLS 3
+	param set BAT_N_CELLS 4
```
Why?

MaEtUgR replied:
1. The `BatterySimulator` is hardcoding this value now so it's what we're already using in SITL atm: https://github.com/PX4/Firmware/blob/4088c2581f38333a306d3cd55aa66b617a51eb93/src/modules/simulator/battery_simulator/BatterySimulator.hpp#L88
1. IMHO 4S is far more common than 3S these days which was not the case when this was originally set up.
1. It doesn't matter, the simulated battery generates a voltage that considers the cell count so voltages "just look more familiar" 🤷‍♂️

dagar commented on src/lib/battery/battery.h:
```diff
@@ -230,6 +214,5 @@ class Battery : public ModuleParams
 
-	orb_advert_t _orb_advert{nullptr};
-	int _orb_instance;
+	orb_advert_t _battery_status_pub{nullptr};
```
```suggestion
	uORB::PublicationMulti<battery_status_s> _battery_status_pub{ORB_ID(battery_status)};
```

MaEtUgR commented:
- fixed the include
- removed unnecessary parameter_update struct
- reworded some things

Diff: https://github.com/PX4/Firmware/compare/e7c302c4c9eb0122864b10b5310257b39ea5e3cd..9770e675c982149e1cf2a79f9a83726e78380e9b

MaEtUgR commented:
I added a commit to switch to `PublicationMulti` like suggested in https://github.com/PX4/Firmware/pull/15616#discussion_r476706755 and https://github.com/PX4/Firmware/pull/15616#discussion_r476909180

julianoes commented on src/modules/battery_status/analog_battery.cpp:
```diff
@@ -1,3 +1,37 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2019 PX4 Development Team. All rights reserved.
```
```suggestion
 *   Copyright (c) 2020 PX4 Development Team. All rights reserved.
```

MaEtUgR replied:
It's from 2019, he just forgot to add the notice 😛

MaEtUgR replied:
I made it 2019-2020 since there are new changes as well.

julianoes commented on src/lib/battery/battery.h:
```diff
@@ -157,67 +155,55 @@ class Battery : public ModuleParams
 	 */
 	template<typename T>
-	bool migrateParam(param_t old_param, param_t new_param, T *old_val, T *new_val, bool firstcall)
+	void migrateParam(param_t old_param, param_t new_param, T *old_val, T *new_val, bool firstcall)
```
Is the comment above still correct?

MaEtUgR replied:
No it isn't ☝️

MaEtUgR replied:
I corrected the description.

julianoes commented on src/lib/battery/battery.h:
```diff
@@ -157,67 +155,55 @@ class Battery : public ModuleParams
-			if ((float) fabs((float) *old_val - (float) *new_val) > FLT_EPSILON) {
-				param_set_no_notification(new_param, old_val);
-				param_get(new_param, new_val);
-				return true;
```
We no longer migrate this direction? How do we keep the old BAT_ params in sync for QGC?

MaEtUgR replied:
We do sync back: See case `!isFloatEqual(*new_val, previous_new_val)` above -> the new value changed
in that case we `param_set_no_notification(old_param, new_val);` save the new changed value to the old parameter handle.

The line you're commenting on is removed because it handles `firstcall == true` for which case the rest of the conditions and action was duplicate. It still does the same thing just above without duplication: `if (firstcall || ...`.

julianoes replied:
Ok good.

MaEtUgR commented:
@julianoes Thanks for the review! I hope I addressed the comments to your liking. If you want to discuss migration cases I'm happy to do so.

dagar commented:
Looks good to me. Let's do an extra round of manual testing here because I'm going to cherry-pick this as the last change for v1.11.

MaEtUgR commented:
@dagar Thanks for reviewing!
Release port like discussed: https://github.com/PX4/Firmware/pull/15654
